src/pytorch_lightning/utilities/data.py

```diff
@@ -94,41 +87,39 @@ def extract_batch_size(batch: BType) -> int:
 
 
 def has_len_all_ranks(
-    dataloader: Union[DataLoader, CombinedLoader],
+    dataloader: object,
     strategy: "pl.strategies.Strategy",
     model: Union["pl.LightningModule", "pl.LightningDataModule"],
-) -> bool:
-    """Checks if a given Dataloader has ``__len__`` method implemented i.e. if it is a finite dataloader or
-    infinite dataloader."""
-    try:
-        local_length = len(dataloader)  # type: ignore [arg-type] # we are checking with duck-typing
```

Comment:
I believe this solves a deadlock that could happen if one rank does not define `__len__`. That rank would jump to the `except` block while the others would try to `reduce`

At least now it raises an exception in that case

Reply:
@carmocca If len is not defined on one of the ranks, it would mean we would be running with different source code. But that's not possible and violates the assumptions of how a program is run.

Reply:
This is cherry-picked, but the following is possible as this is a public function

```python
thing = SizedThing()
if rank == 1:
   thing = UnsizedThing()
has_len_all_ranks(thing, ...)
```

In any case, this was a minor observation. As you suggest, it shouldn't happen within our own usage.